### PR TITLE
Added cluster validator webhook with unique local cluster CR validation

### PR DESCRIPTION
**Pull request by pregnor**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | STB-554
| License         | Apache 2.0

### What's in this PR?

1. Implemented a self signed certificate generation method for webhooks.
2. Implemented a validating admission webhook for cluster CR validations.
3. Also implemented a validation for ensuring unique local cluster CRs based on cluster type (local) and cluster ID (unique).

**By default the cluster CR validator webhook is turned on through Helm chart values.** Let me know if it should be turned off by default.

### Why?

To disallow creating cluster CRs with cluster ID already belonging to an existing local typed cluster CR.
This is an improvement for multi-cluster SDM setups.

### Additional context

I based most of the implementation on the banzaicloud/istio-operator#525 PR.

Tested manually with the following scenarios:
1. Allowed creating local cluster CR successfully.

```yaml
# Creating local cluster CR is successful.
kind: Cluster
apiVersion: clusterregistry.k8s.cisco.com/v1alpha1
metadata:
  name: pregnor-1
spec:
  authInfo:
    secretRef:
      name: pregnor-1
      namespace: cluster-registry
  clusterID: 80feb07e-3c69-4b98-a576-f62d5df65566
  kubernetesApiEndpoints:
  - serverAddress: https://3.72.141.112:6444
```

2. Allowed updating local cluster CR with identical cluster ID and name compared to existing local cluster CR.

```yaml
# Updating existing local cluster CR with identical cluster ID and name is successful.
kind: Cluster
apiVersion: clusterregistry.k8s.cisco.com/v1alpha1
metadata:
  name: pregnor-1
spec:
  authInfo:
    secretRef:
      name: pregnor-1
      namespace: cluster-registry
  clusterID: 80feb07e-3c69-4b98-a576-f62d5df65566
  kubernetesApiEndpoints:
  - serverAddress: https://3.72.141.112:6443
```

3. Rejected creating local cluster CR with identical cluster ID and different name compared to existing local cluster CR.

```yaml
# Creating local cluster CR with identical cluster ID and different name than existing local cluster CR is rejected.
kind: Cluster
apiVersion: clusterregistry.k8s.cisco.com/v1alpha1
metadata:
  name: pregnor-1-2
spec:
  authInfo:
    secretRef:
      name: pregnor-1-2
      namespace: cluster-registry
  clusterID: 80feb07e-3c69-4b98-a576-f62d5df65566
  kubernetesApiEndpoints:
  - serverAddress: https://3.72.141.112:6443
```

```log
Error from server: error when creating "test_3_failure.yaml": admission webhook "cluster-validator.clusterregistry.k8s.cisco.com" denied the request: validating unique local cluster CR failed: a local cluster CR with name pregnor-1 already exists with the same clusterID 80feb07e-3c69-4b98-a576-f62d5df65566
```

4. Allowed creating peer cluster CR with different cluster ID and name compared to existing local cluster CR.

```yaml
# Creating peer cluster CR with different cluster ID and name compared to existing local cluster CR is successful.
kind: Cluster
apiVersion: clusterregistry.k8s.cisco.com/v1alpha1
metadata:
  name: pregnor-2
spec:
  authInfo:
    secretRef:
      name: pregnor-2
      namespace: cluster-registry
  clusterID: 7aa85423-31c6-4029-98fc-7fff3e440693
  kubernetesApiEndpoints:
  - serverAddress: https://3.70.12.109:6443
```

5. Replaced the `ValidatingWebhookConfiguration` with a similar `MutatingWebhookConfiguration` to see if the webhook certifier works with mutating webhooks as well.

### Checklist

- [X] Implementation tested
- ~User guide and development docs updated (if needed)~

### To Do

- [x] todo(@pregnor): clean up unneeded istio(-operator) references and unneeded transported code.

### CODE REVIEW

**Comment by pregnor** [comment]
CC: @Laci21 , @waynz0r  (as I cannot add reviewers)

**Comment by pregnor** [comment]
To be cleaned up, there were a couple mistaken cross-references left in from code taken from earlier implementations.

File: go.mod [hunk 1]
```diff
@@ -27,6 +27,14 @@ require (
 	sigs.k8s.io/yaml v1.2.0
 )
 
+require (
+	github.com/Masterminds/semver/v3 v3.1.1
+	github.com/banzaicloud/istio-operator/pkg/apis v0.10.8
+	github.com/goph/emperror v0.17.2
+	golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
+	k8s.io/apiextensions-apiserver v0.21.3
```
Comment (stoader) [inline go.mod:35]: Can we use version `v0.23.1`?

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/cert/certificate_provisioner.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+func (p *CertificateProvisioner) generateCert(dnsNames []string) (*Certificate, error) {
+	var certificate Certificate
+	var caPEM, caPrivateKeyPEM, serverCertPEM, serverPrivKeyPEM *bytes.Buffer
+	// CA config
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(2020),
+		Subject: pkix.Name{
+			Organization: []string{"istio-operator"},
```
Comment (stoader) [inline pkg/cert/certificate_provisioner.go:139]: Can't we just set common name instead of organization and set to `cluster-registry-controller-ca`?

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/cert/certificate_provisioner.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		commonName = dnsNames[0]
+	}
+
+	// server cert config
+	cert := &x509.Certificate{
+		DNSNames:     dnsNames,
+		SerialNumber: big.NewInt(1658),
+		Subject: pkix.Name{
+			CommonName:   commonName,
+			Organization: []string{"istio-operator"},
```
Comment (stoader) [inline pkg/cert/certificate_provisioner.go:187]: I believe organization can be omitted of common name is provided.

File: pkg/k8sutil/desired_state.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/cisco-open/cluster-registry-controller/pkg/k8sutil/wait"
+	"github.com/cisco-open/cluster-registry-controller/pkg/util"
+)
+
+const (
+	detachedPodLabel = "istio.banzaicloud.io/detached-pod"
```
Comment (stoader) [inline pkg/k8sutil/desired_state.go:35]: The label domain should be in line with cluster registry (e.g. `clusterregistry.k8s.cisco.com/detached-pod`)

File: pkg/k8sutil/predicates.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	istiov1beta1 "github.com/banzaicloud/istio-operator/pkg/apis/istio/v1beta1"
+)
+
+func GetWatchPredicateForIstio() predicate.Funcs {
```
Comment (stoader) [inline pkg/k8sutil/predicates.go:29]: Cluster registry controller should not watch istio-operator related types.

**Review by stoader — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (pregnor) [reply go.mod]: Fixed in https://github.com/cisco-open/cluster-registry-controller/compare/a0fb60a438bc2210e4dae9bc4992473b00263051..a4c0ccb10620a661c53c92315574dca2d72d4a99 .

Removed the code directly depending on it, it remained as an indirect dependency used by the cluster-registry-controller.

File: pkg/cert/certificate_provisioner.go (hunk 3 above)
Comment (pregnor) [reply pkg/cert/certificate_provisioner.go]: Fixed in https://github.com/cisco-open/cluster-registry-controller/compare/a0fb60a438bc2210e4dae9bc4992473b00263051..a4c0ccb10620a661c53c92315574dca2d72d4a99 .

Removed the organization.

File: pkg/k8sutil/desired_state.go (hunk 4 above)
Comment (pregnor) [reply pkg/k8sutil/desired_state.go]: Fixed in https://github.com/cisco-open/cluster-registry-controller/compare/a0fb60a438bc2210e4dae9bc4992473b00263051..a4c0ccb10620a661c53c92315574dca2d72d4a99 .

Removed this code.

**Review by Laci21 — APPROVED** [review]

File: deploy/charts/cluster-registry/templates/service.yaml [hunk 6]
```diff
@@ -12,5 +12,11 @@ spec:
       targetPort: metrics
       protocol: TCP
       name: http-metrics
+    {{- if .Values.webhooks.clusterValidator.enabled }}
+    - port: 443
+      targetPort: cl-val-wh
+      protocol: TCP
+      name: cl-val-wh
```
Comment (waynz0r) [inline deploy/charts/cluster-registry/templates/service.yaml:19]: can we use http-cl-val-wh as a name instead?

**Review by waynz0r — CHANGES_REQUESTED** [review]

File: deploy/charts/cluster-registry/templates/service.yaml (hunk 6 above)
Comment (pregnor) [reply deploy/charts/cluster-registry/templates/service.yaml]: Fixed in https://github.com/cisco-open/cluster-registry-controller/pull/22/commits/aa3f05ffe506a32c991eb4223f0ba44c18e5e2f2

**Review by waynz0r — APPROVED** [review]

**Pull request merged**